### PR TITLE
feat(code-quality): add golangci-lint to CI and lint project

### PR DESCRIPTION
**Pull request by darkweak**
Hello, this PR is here to have a better code quality/code style using `golangci-lint`. It adds the CI job to ensure every other PR will be as well written as possible.

### CODE REVIEW

File: .github/workflows/go_generate_update.yml [hunk 1]
```diff
@@ -35,6 +35,9 @@ jobs:
                 run: |
                     RESULT=$(git status --untracked-files=no --porcelain)
                     echo "gitstatus=$RESULT" >> $GITHUB_OUTPUT
+            - 
+                name: golangci-lint
+                uses: golangci/golangci-lint-action@v3
```
[the remark this answers is not in the document]
Comment (tucksaun) [reply .github/workflows/go_generate_update.yml:40]: we could enable `only-new-issues` in order to not overwhelm new contributors with code not up to date with quality

**Review by tucksaun — CHANGES_REQUESTED** [review]
Thank you @darkweak for working on this 🙏
This has been on my mind for a while 👍

Just a couple of thoughts:

- I believe we should tweak the settings to match our use case though.
For example, `unusedresult` should probably be disabled (at least for several functions from `terminal`).

- We should also take the opportunity to some of those issues instead of masking them (not checking errors for example) which makes me believe it would be better to add the configuration in this PR, maybe fix the obvious issues (`ioutil`-> `os` for example) and fix the other issues with other PRs.
 
any thought @fabpot ?

File: .github/workflows/go_generate_update.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (darkweak) [reply .github/workflows/go_generate_update.yml:40]: You're right ✌️

**Comment by darkweak** [comment]
> I believe we should tweak the settings to match our use case though.
For example, unusedresult should probably be disabled (at least for several functions from terminal).

IMHO if that's unused that should be removed 😄 

> We should also take the opportunity to some of those issues instead of masking them

That's the first iteration and was to implement asap the golangci-lint. I plan to add some error handling (for those I just disabled checks).

File: .github/workflows/go_generate_update.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (darkweak) [reply .github/workflows/go_generate_update.yml:40]: After thinking about that, 
> we could enable only-new-issues in order to not overwhelm new contributors with code not up to date with quality

If they rebase their PR they'll have the correct code.

**Comment by tucksaun** [comment]
> IMHO if that's unused that should be removed 😄

I'm not referring to unused but unused results that are not always useful and I personally find having tons of `_, _  = ` unreadable (see all the changes for functions as `terminal.Printfln`)

> That's the first iteration and was to implement asap the golangci-lint. I plan to add some error handling (for those I just disabled checks).

That is the reason why I mentioned enabling only-new-issues. By doing so we can more quickly merge this PR that would only add configuration and most obvious fix and then we can fix other issues in dedicated PRs

**Comment by darkweak** [comment]
Oh sorry, I misunderstood what you said then, thank you for your explanations.
I'll try to find time to update the PR during this week-end 👍

**Comment by darkweak** [comment]
@tucksaun the PR is ready.

**Comment by darkweak** [comment]
@fabpot @tucksaun any update?

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  timeout: 30s
+  issues-exit-code: 1
+
+linters:
+  disable:
+    # TODO remove these linters and fix the deadcode/errcheck/unused/varcheck
+    - deadcode
+    - errcheck
+    - unused
+    - varcheck
```
Comment (tucksaun) [inline .golangci.yml:11]: we can remove `deadcode`, `varcheck` and `unused` from the list

Comment (tucksaun) [reply .golangci.yml:11]: also missing new line at the end of the file

File: book/checkout.go [hunk 3]
```diff
@@ -35,7 +35,9 @@ import (
 func (b *Book) Checkout(step string) error {
 	// FIXME: keep vendor/ node_modules/ around before git clean, but them back as they will be updated the right way, less Internet traffic
 	// FIXME: if the checkout is to a later step, no need to remove the DB, we can just migrate it
-	os.Chdir(b.Dir)
+	if err := os.Chdir(b.Dir); err != nil {
+		return err
```
Comment (tucksaun) [inline book/checkout.go:39]: We usually wrap errors (with `errors.WithStack`) to get nice stack traces

Comment (tucksaun) [reply book/checkout.go:39]: this also means `wrapcheck` should probably be enabled in golangci configuration

File: .golangci.yml (hunk 2 above)
Comment (tucksaun) [reply .golangci.yml:11]: and regarding errcheck, given the fact that you fixed most of then we can enable errcheck and disable it only for terminal functions like `Println`

Comment (tucksaun) [reply .golangci.yml:11]: (see my other configuration suggestion)

Comment (tucksaun) [inline .golangci.yml:11]:
```suggestion
linters:
  enable:
    - gci
    - wrapcheck

linters-settings:
  wrapcheck:
    ignorePackageGlobs:
      # We already make sure your own packages wrap errors properly
      - github.com/symfony-cli/*
  errcheck:
      ignore: github.com/symfony-cli/terminal:^(Ep|P)rint
```

**Review by tucksaun — CHANGES_REQUESTED** [review]

**Comment by darkweak** [comment]
Ping @tucksaun

File: .golangci.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    - gci
+    - wrapcheck
+
+linters-settings:
+  wrapcheck:
+    ignorePackageGlobs:
+      # We already make sure your own packages wrap errors properly
+      - github.com/symfony-cli/*
+  errcheck:
+      ignore: github.com/symfony-cli/terminal:^(Ep|P)rint
```
Comment (tucksaun) [inline .golangci.yml:16]: missing newline at the end of the file

File: book/checkout.go [hunk 5]
```diff
@@ -35,7 +35,9 @@ import (
 func (b *Book) Checkout(step string) error {
 	// FIXME: keep vendor/ node_modules/ around before git clean, but them back as they will be updated the right way, less Internet traffic
 	// FIXME: if the checkout is to a later step, no need to remove the DB, we can just migrate it
-	os.Chdir(b.Dir)
+	if err := os.Chdir(b.Dir); err != nil {
+		return errors.WithStack(err)
+	}
```
Comment (tucksaun) [inline book/checkout.go:40]:
```suggestion
	if err := errors.WithStack(os.Chdir(b.Dir)); err != nil {
		return err
	}
```

I'm not saying we have to do it, but we can do it when there's only a single return value, this also better preserve stack traces IIRC

File: book/checkout.go [hunk 6]
```diff
@@ -110,40 +112,44 @@ func (b *Book) Checkout(step string) error {
...
 		if err := executeCommand(append(dockerComposeBin(), "down", "--remove-orphans"), b.Debug, false, nil); err != nil {
-			return err
+			return errors.WithStack(err)
 		}
 	} else {
 		terminal.Println("Skipped for this step")
 	}
 
 	printBanner("<comment>[WEB]</> Stopping the Local Web Server", b.Debug)
-	executeCommand([]string{"symfony", "server:stop"}, b.Debug, true, nil)
```
Comment (tucksaun) [inline book/checkout.go:120]: I would tend to think that @fabpot deliberately skip this one

File: book/checkout.go [hunk 7]
```diff
@@ -110,40 +112,44 @@ func (b *Book) Checkout(step string) error {
...
 
 	printBanner("<comment>[GIT]</> Checking out the step", b.Debug)
 	if err := executeCommand([]string{"git", "checkout", "-B", branch, tag}, b.Debug, false, nil); err != nil {
-		return err
+		return errors.WithStack(err)
 	}
 
 	printBanner("<comment>[SPA]</> Stopping the Local Web Server", b.Debug)
 	if _, err := os.Stat(filepath.Join(b.Dir, "spa")); err == nil {
-		executeCommand([]string{"symfony", "server:stop", "--dir", filepath.Join(b.Dir, "spa")}, b.Debug, true, nil)
```
Comment (tucksaun) [inline book/checkout.go:134]: same

File: commands/local_proxy_domain_attach.go [hunk 8]
```diff
@@ -40,16 +41,16 @@ var localProxyAttachDomainCmd = &console.Command{
 	Action: func(c *console.Context) error {
 		projectDir, err := getProjectDir(c.String("dir"))
 		if err != nil {
-			return err
+			return errors.WithStack(err)
```
Comment (tucksaun) [inline commands/local_proxy_domain_attach.go:44]: this one should be wrapped in the parent function

File: commands/local_proxy_domain_attach.go [hunk 9]
```diff
@@ -40,16 +41,16 @@ var localProxyAttachDomainCmd = &console.Command{
...
 		}
 
 		homeDir := util.GetHomeDir()
 		config, err := proxy.Load(homeDir)
 		if err != nil {
-			return err
+			return errors.WithStack(err)
```
Comment (tucksaun) [inline commands/local_proxy_domain_attach.go:50]: same

File: commands/local_proxy_start.go [hunk 10]
```diff
@@ -68,7 +68,7 @@ var localProxyStartCmd = &console.Command{
 			}
 			if err := reexec.Background(varDir); err != nil {
 				if _, isExitCoder := err.(console.ExitCoder); isExitCoder {
-					return err
+					return errors.WithStack(err)
```
Comment (tucksaun) [inline commands/local_proxy_start.go:71]: IIRC `console.ExitCoder` errors must not be wrapped as we have special handling for them in `symfony-cli/console`

**Review by tucksaun — CHANGES_REQUESTED** [review]
I'm stopping here because reviewing this kind of PR is complicated 😖.
Also, the PR needs rebasing.
Would you reconsider opening a new PR with only the configuration and then creating one PR per topic fixed (one by one)?
That would help a lot IMO as they are a couple of fixes that could be merged quickly then.

**Comment by darkweak** [comment]
@tucksaun I'll open some PRs to review easier. I'll start with one that add only golangci-lint in the project with the given configuration without noises. The PR should be ready this evening. ✌️

**Pull request closed without merging**